Add 'Upload Files...' menu entries

akosyakov opened the pull request:
fix #4088

TODO:
- [x] fix handling of global selection by the navigator: it does not clean it up on disposal
- [x] relocate & group download & upload in the main menu
- [x] relocate & group download & upload in the navigator context menu
- [x] reveal a folder in the navigator after upload 
- [x] ~open a CQ for~ formidable - it's MIT and does not have 3rd party dependencies, we can wait till the CQ process for prod dependencies is clear
- [x] sign off a commit properly
- [x] uploading to the same folder twice should work

![upload](https://user-images.githubusercontent.com/3082655/55057375-52a76100-5069-11e9-91ee-ef14e21c4d48.gif)

akosyakov commented on packages/filesystem/src/browser/download/file-download-command-contribution.ts:
```diff
@@ -57,29 +76,20 @@ export class FileDownloadCommandContribution implements CommandContribution {
         return this.isDownloadEnabled(uris);
     }
 
-    protected getUris(uri: Object | undefined): URI[] {
```
@kittaakos i've removed it, seems to be just left overs which were never used

akosyakov commented:
@jankeromnes @jgbradley1 Could you try it out?

akosyakov commented:
@vzhukovskii There was a bug in the navigator and plugin tree views that the global selection was not cleaned properly when they are disposed. I've introduced `TreeWidgetSelection` to deal with that in `TreeWidget`. One can enable it via `globalSelection` props. Could you please have a look that it does not break menu items for tree view? I would it as well.

vzhukovskii commented:
@akosyakov thanks! Will try to check it asap.

vzhukovskii commented:
Looks like it doesn't break existed logic on plugin side.

jgbradley1 commented:
Nice. I like it! One minor UI issue. If a user has a directory open (i.e. by clicking on the drop down arrow) in the workspace and uploads files to it, they will not actually see the file structure updated with the new file names. You must first close the folder and reopen it.

Is there a function call to refresh the file tree after the upload process is complete? This same issue exist when you delete a file. You have to close the folder and open it back up again to see the changes occur.

jgbradley1 commented:
This is not a necessary feature right now, but we should think about some way to communicate upload progress on a status bar or on the blue bar at the bottom of the webpage.

akosyakov commented:
> This is not a necessary feature right now, but we should think about some way to communicate upload progress on a status bar or on the blue bar at the bottom of the page.

It is on our list to allow blueish progress bar for different top-level widgets similar what you see in VS Code or Chrome while loading the page. It is even more important for cloud solution than for VS Code, since connection delay and issues is something which we need to deal with.

akosyakov commented:
>  If a user has a directory open (i.e. by clicking on the drop down arrow) in the workspace and uploads files to it, they will not actually see the file structure updated with the new file names. You must first close the folder and reopen it.

It is fs watching issue. I will have a look into it. A library which we use has troubles with symlinks and does not report some events for real paths :( Refreshing for each operation is work around, it should happen automatically based on fs events.

akosyakov commented:
- [ ] found a bug that uploading to the same folder twice does not work

akosyakov commented:
@jgbradley1 could you try again, i've added a fix for fs watching with symlinks

AlexTugarev commented on packages/filesystem/src/node/nsfw-watcher/nsfw-filesystem-watcher.ts:
```diff
@@ -192,6 +192,28 @@ export class NsfwFileSystemWatcherServer implements FileSystemWatcherServer {
+            }
+        }
+    }
+    protected realpath(path: string): string {
```
left over?

AlexTugarev commented on packages/filesystem/src/node/nsfw-watcher/nsfw-filesystem-watcher.ts:
```diff
@@ -111,17 +111,17 @@ export class NsfwFileSystemWatcherServer implements FileSystemWatcherServer {
                 }
                 if (event.action === nsfw.actions.MODIFIED) {
-                    this.pushUpdated(watcherId, paths.join(event.directory, event.file!));
+                    this.pushUpdated(watcherId, this.resolvePath(event.directory, event.file!));
```
Q: shouldn't both be reported as changed? I mean the link and the target.

akosyakov replied:
i would need to check whether it is possible to report for all symlins, if not then we stick only with the real part. reportin fs events should be fast, detecting all symlinks would harm perf

AlexTugarev replied:
I checked with vscode behavior, and this change would make it quite similar.

AlexTugarev commented on packages/filesystem/package.json:
```diff
@@ -6,6 +6,7 @@
     "@theia/core": "^0.4.0",
     "@types/base64-js": "^1.2.5",
     "@types/body-parser": "^1.17.0",
+    "@types/formidable": "^1.0.31",
```
do we need a CQ for this?

AlexTugarev replied:
verified it's MIT. thanks!

akosyakov commented:
i think we can wait till someone opens bugs for pseudo files and silent override. What should be good default `maxFileSize`?

akosyakov commented:
@AlexTugarev I've added proper error handling and `uploadMaxFileSize` env variable, by default it is 2048MB. Could you check again?

@jgbradley1 i've also added upload progress reporting :)

elaihau commented:
Nice feature. i love it. 
I tried uploading different versions of the same file to the same location, and found the newer version replaces the older - that's what i expected. would it be helpful if theia warns the user the older version would be overwritten ?

akosyakov commented:
> I tried uploading different versions of the same file to the same location, and found the newer version replaces the older - that's what i expected. would it be helpful if theia warns the user the older version would be overwritten ?

it is not easy to implement, this part of code cannot provide responses to a user, we can only fail the whole request.

jgbradley1 commented:
@akosyakov This is great. +1 on the progress bar. Something is wrong with the max file size. It's off by an order of magnitude. I cannot upload any file > 2 MB and they all fail with the `Payload too large` message (tested on Gitpod). Files < 2 MB upload successfully.  The math in your code looks correct however.

`2048 * 1024 * 1024 (bytes) = 2 GB`

akosyakov commented:
@jgbradley1 good catch! thank you for checking

akosyakov commented:
@AlexTugarev @jgbradley1 I've fixed and retested locally handling of exceeded payload. It should be fine now. Good to merge?

akosyakov commented:
merging, @AlexTugarev said that he is fine with changes in off-line conversation
